### PR TITLE
NextIP should properly convert IPv4 and IPv6 with leading 0

### PR DESCRIPTION
**Pull request by alexanderConstantinescu**
**- What this PR does and why is it needed**

During my implementation of EgressIP I noticied incorrectly computed next IPs for IP addresses with leading 0s. This fixes that

/cc @danwinship 

**- Special notes for reviewers**

**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

File: go-controller/pkg/util/net.go [hunk 1]
```diff
@@ -12,19 +12,19 @@ import (
...
+	byteSize := 0
+	v := ip.To4()
+	if v != nil {
+		byteSize = 4
+	} else {
+		v = ip.To16()
+		byteSize = 16
+	}
+	i := big.NewInt(0).SetBytes(v)
+	i.Add(i, big.NewInt(1))
```
Comment (gvbalaji) [inline go-controller/pkg/util/net.go:24]: What happens in case of edge conditions such as 254?

**Comment by danwinship** [comment]
> What happens in case of edge conditions such as 254? I meant i test case on edge condition would be helpful.

Right, so the answer is, "if you hit the edge condition, you lose". The code doesn't know where `ip`'s subnet ends, so it can't do anything sane on overflow, because it can't tell when it has overflowed. Basically, you are only allowed to call this function if you know it's _not_ going to overflow.

But then, that means that all the stuff with `"math/big"` is pointless; we can just increment the low byte of the `net.IP`, and if it doesn't overflow, we get the same answer as the current code would, and if it does overflow, then the answer is probably wrong anyway.

**Comment by girishmg** [comment]
@danwinship An interesting question is what will happen when we do

```
NextIP(net.ParseIP("255.255.255.255"))
or
NextIP(net.ParseIP("ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff"))
```

the code in this PR results in a `panic`.

```
panic: runtime error: slice bounds out of range [-1:]

goroutine 1 [running]:
main.NextIP(0xc0000ae240, 0x10, 0x10, 0x10, 0x10, 0x12)
	/tmp/sandbox287422424/prog.go:22 +0x487
main.main()
	/tmp/sandbox287422424/prog.go:42 +0xce
```

One could argue, that we never should hit this panic. However, it is fairly easy fix if we check for the length of bytes after we add `1`. So,

```
i.Add(i, big.NewInt(1))
blen := len(i.Bytes())
if byteSize == 4 && blen > 4 {
   return nil
} else if byteSize == 16 && blen > 16 {
  return nil
}
```

**Comment by gvbalaji** [comment]
I think as @danwinship suggested, taking just the last byte and incrementing that would be simpler. We don't need BigInt arithmetic here.  I am thinking more in those lines:
1. Take the last byte
2. If it's >=254 return nil (error). This way we will check the absolute boundary condition for the legal IP address. 
3. Else increment it and send the original byte array with modified last byte.

**Comment by girishmg** [comment]
That will not help. If the subnet, is 192.168.0.0/16 and the current IP is 192.168.0.255. The NextIP should return 192.168.1.0 and not error.

Either we fix NextIP further by making it aware of subnet boundaries or have the current fix in this PR plus the fix to avoid the panicI and leave it upto the callers for checking on subnet boundaries overflow.

**Comment by danwinship** [comment]
> That will not help. If the subnet, is 192.168.0.0/16 and the current IP is 192.168.0.255. The NextIP should return 192.168.1.0 and not error like you mention. The code in this PR and the current code correctly returns the NextIP. The BigInt math is required.

But the current IP _isn't_ 192.168.0.255. We only ever call `NextIP` on the first or second IP in a subnet, and we don't allow /31 subnets.

> Either we fix the NextIP further by making it aware of subnet boundaries

We could trivially replace all existing uses of `NextIP` with `GetNthIP(subnet *net.IPnet, n int) net.IP`.

But we should make explicit the current implicit semantics that callers don't have to check for an error/`nil` return, because it can't fail if you called it with valid arguments (eg, any supported OVN-Kubernetes subnet, and n <= 2)

**Comment by girishmg** [comment]
> We could trivially replace all existing uses of `NextIP` with `GetNthIP(subnet *net.IPnet, n int) net.IP`.
> 
> But we should make explicit the current implicit semantics that callers don't have to check for an error/`nil` return, because it can't fail if you called it with valid arguments (eg, any supported OVN-Kubernetes subnet, and n <= 2)

Yes, this will work.

**Comment by alexanderConstantinescu** [comment]
>We only ever call `NextIP` on the first or second IP in a subnet, and we don't allow /31 subnets.

So, I think I'm going to close this PR unless I hear any objection from you guys.  

The only reason I ever needed this patch was because I had implemented an IP allocation manager for the assignment of EgressIP. Now that I see https://github.com/ovn-org/ovn-kubernetes/pull/1374 merged, I am going to throw away my code and use that instead. And given the above quote, it does not seem like this patch is needed given the operations we perform when using `NextIP`

**Comment by girishmg** [comment]
see issue #1386
